src/OpenTelemetry/Metrics/MetricReader.cs

```diff
@@ -59,8 +61,12 @@ public virtual void OnCollect(Batch<Metric> metrics)
         /// <inheritdoc/>
         public void Dispose()
         {
-            this.Dispose(true);
-            GC.SuppressFinalize(this);
+            if (!this.disposed)
+            {
+                this.Dispose(true);
+                GC.SuppressFinalize(this);
+                this.disposed = true;
+            }
```

Comment:
I think we don't have native resource so Finalizer is not needed (and having it will cause extra overhead)?

Or do you think we want to allow derived class to have native resource, and prevent folks from leaking them if they don't dispose the instances properly?

Reply:
> Or do you think we want to allow derived class to have native resource, and prevent folks from leaking them if they don't dispose the instances properly?

Exactly! We can't really control what users will do with it. Only time I think it is safe to leave out finalizer is if class is `internal` or `sealed` and not using any non-managed resources. I don't think there is a major perf hit for users correctly calling `Dispose` do to the `GC.SuppressFinalize(this);` call, but I haven't measured it.

Reply:
Just looking at the [docs](https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-dispose), I think my initial comment was messed up! Sorry. Looks like `MetricReader` should just be...

```csharp
public abstract class MetricReader : IDisposable
{
    public void Dispose()
    {
        this.Dispose(true);
        GC.SuppressFinalize(this);
    }

    protected virtual void Dispose(bool isDisposing)
    {
    }
}
```

Derived classes would be...

```csharp
public class MyMetricReader : MetricReader 
{
    private readonly SemaphoreSlim semaphore = new();
    private bool disposed;

    protected override void Dispose(bool isDisposing)
    {
        if (this.disposed)
        {
            return;
        }

        if (disposing)
        {
            this.semaphore.Dispose();
        }

        this.disposed = true;

        base.Dispose(disposing);
    }
}

public class MyMetricReaderWithNative : MetricReader 
{
    private IntPtr handle;
    private bool disposed;

    ~MyMetricReaderWithNative() => this.Dispose(false);

    protected override void Dispose(bool isDisposing)
    {
        if (this.disposed)
        {
            return;
        }

        // todo: clean up handle regardless of isDisposing

        this.disposed = true;

        base.Dispose(disposing);
    }
}
```

Reply:
Related to #1024.

Reply:
Done.